### PR TITLE
feat: improve pilotnet prepare_data with shuffle and frame decimation

### PR DESCRIPTION
**Pull request by iwatake2222**
## 変更内容

- PilotNet学習のヘルパースクリプトであるprepare_data.pyに以下の変更を行います
  - データのシャッフル
    - trainとvalに振り分ける前にデータの全フレームをシャッフルし、配列の前半と後半でtrainとvalに振り分けます
  - データの間引き用オプションの追加（ `--frame-stride` 。デフォルト=1 = 間引きなし）

## 変更理由

- シャッフルが必要な理由
  - 現状は、all-dirフォルダ直下の各フォルダを順次読み込み、前半80%と後半20%でtrainとvalに分けています。しかしフォルダ単位で前半と後半に分けられると、後半のフォルダのみに入っているシーンの学習が行われません。そのためシャッフルが必要です
- データの間引きが必要な理由
  - 現状は、全フレームのデータを学習に使用しています。しかし、隣接するフレームの入出力は通常は酷似しており冗長です。（通常の物体検出でも1秒ごとに間引くことが多いです）。加えて、上述のシャッフルによって隣接した酷似するフレームがtrainとvalに振り分けられるリスクがあります（データのリークの可能性）
  - その一方で、十分な学習データを集められない可能性もあります。その場合にはたとえ酷似していても、隣接フレームを学習に含めたほうがマシです。そのため、デフォルトは間引きなしのframe-stride=1にしています。

### 変更前

valに使用される後半のフォルダのシーンが学習に含まれていないため、trainのlossは下がりますが、valのlossは下がりません

image

### 変更後

valのlossも下がるようになりました。

image

## 備考

PilotNetは時系列情報は見ないため、シャッフルして順番が変わっても問題ありません

### CODE REVIEW

File: aichallenge/ml_workspace/pilot_net/prepare_data.py [hunk 1]
```diff
@@ -46,8 +48,15 @@ def main(all_dir: Path, train_dir: Path, val_dir: Path, split_ratio: float = 0.8
     del all_imgs, all_steers, all_accels
 
     n = len(steers)
+
+    # --- Shuffle and split (decimation above already removed most leakage) ---
+    perm = np.random.permutation(n)
```
[the remark this answers is not in the document]
Comment (iwatake2222) [reply aichallenge/ml_workspace/pilot_net/prepare_data.py:53]: Addressed in [e9a54d9](https://github.com/AutomotiveAIChallenge/aichallenge-racingkart/pull/262/commits/e9a54d90e16470a4567e0097647cbe72d02cd271)

Thanks — this is a valid point, and we agree the default --frame-stride 1 leaves temporal leakage in place (near-duplicate adjacent frames can land in both train and val, and train.py uses val loss for checkpointing, LR scheduling, and early stopping).

  This is an intentional trade-off. The input is ~10 fps, so a stride of 5–10 (0.5–1 s spacing) is the appropriate setting once enough data is available, and we recommend it in those cases. We keep the default at 1 (no decimation) so the script works out of the box for users who
  can't collect much training data — for them, including near-duplicate frames is preferable to discarding data.

  We've documented this in the README (recommend --frame-stride 5–10 when data is sufficient, default 1 otherwise) so the leakage caveat and the recommended value are explicit. Sequence-/block-level splitting is a cleaner long-term fix and something we can revisit if needed.

**Review by taikitanaka3 — COMMENTED** [review]
LGTMです！

**Review by taikitanaka3 — APPROVED** [review]
